proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+        value = matches ? value2 : default!;
+        return matches;
+    }
+    
+    object? IUnion.Value => kind switch
+    {
+        Kind.Case1 => value1,
+        Kind.Case2 => value2,
+        _ => null;
+    };
+}
+```
+
+#### Well-formedness
+
+The language and compiler make a number of behavioral assumptions about union types. If a type qualifies as a union type but does not satisfy those assumptions, then union behaviors may not work as expected.
+
+* *Soundness*: The `Value` property always evaluates to null or to a value of a case type.
+* *Stability*: If a union value is constructed from a case type, the `Value` property will match that case type.
+* *Constructor equivalence*: If a value is implicitly convertible to two different case types then the constructor for either of those case types has the same observable behavior when called with that value.
+* *Access pattern consistency*: The behavior of `HasValue` and `TryGetValue` members of the non-boxing access pattern, if present, is observably equivalent to that of checking against the `Value` property directly.
+
+### Union behaviors
+
+The union behaviors are generally implemented by means of the basic union pattern. If the union offers the non-boxing access pattern, union pattern matching will preferentially make use of it.
+
+#### Union conversions
+
+A *union conversion* implicitly converts to a union type from each of its case types. It is sugar for calling the union's constructor for the given case type:
+
+``` c#
+Pet pet = dog;
+// becomes
+Pet pet = new Pet(dog);
+```
+
+It is not an error if more than one constructor overload applies. Instead, one is chosen in an implementation-defined manner.
+
+#### Union matching
+
+When the incoming value of a pattern is of a union type, the union value's contents may be "unwrapped", depending on the pattern.
+
+For the unconditional `_` and `var` patterns, the pattern is applied to the union value itself. For example:
+
+```csharp
+if (GetPet() is var pet) { ... } // 'pet' is the union value returned from `GetPet`
+```
+
+However, all other patterns get implicitly applied to the contents of the union:
+
+``` c#
+if (GetPet() is Dog dog) { ... } // 'Dog dog' is applied to the contents of the Pet value
+if (GetPet() is null) { ... } // 'null' is applied to the contents of the Pet value
+```
+
+If members prescribed by the non-boxing access pattern can be used to implement the pattern behavior, they will. The rules are as follows:
+
+* For a pattern that implies checking for a specific type `T`, if a `TryGetValue(S value)` method is available, and there is an implicit conversion from `T` to `S`, then that method is used to obtain the value. The pattern is then applied to that value. If there is more than one such method, then any where the conversion from `T` to `S` is not a boxing conversion is preferred if available. If there is still more than one method, one is chosen in an implementation-defined manner.
```

Comment:
what's teh value int eh 'implementation defined' parts of this (and one of hte clauses a page back)?  What are the use cases we think this is good for?

Reply:
It's for when case types are overlapping. That goes for interfaces and type parameters, but also actually for every reference type, since they all allow `null`.

We don't want the compiler to go all ambiguous here, since the well-formedness rules ensure that any pick is equally valid and does the same thing.